open3d_slam/src/ScanToMapRegistration.cpp

```diff
@@ -0,0 +1,93 @@
+/*
+ * ScanToMapRegistration.cpp
+ *
+ *  Created on: Oct 31, 2022
+ *      Author: jelavice
+ */
+#include "open3d_slam/ScanToMapRegistration.hpp"
+#include "open3d_slam/Submap.hpp"
+#include "open3d_slam/helpers.hpp"
+#include "open3d_slam/assert.hpp"
+
+#include "open3d_slam/cloud_processing.hpp"
+
+namespace o3d_slam {
+
+namespace {
+namespace registration = open3d::pipelines::registration;
+std::shared_ptr<registration::TransformationEstimation> icpObjective;
+open3d::pipelines::registration::ICPConvergenceCriteria icpCriteria;
+} // namespace
+
+ScanToMapIcpOpen3D::ScanToMapIcpOpen3D() {
+	update(params_);
+}
+
+void ScanToMapIcpOpen3D::setParameters(const MapperParameters &p) {
+	params_ = p;
+	update(params_);
+}
+void ScanToMapIcpOpen3D::update(const MapperParameters &p) {
+	icpCriteria.max_iteration_ = p.scanMatcher_.maxNumIter_;
+	icpObjective = icpObjectiveFactory(p.scanMatcher_.icpObjective_);
+	mapBuilderCropper_ = croppingVolumeFactory(params_.mapBuilder_.cropper_);
+	scanMatcherCropper_ = croppingVolumeFactory(params_.scanProcessing_.cropper_);
+}
+
+void ScanToMapIcpOpen3D::estimateNormalsIfNeeded(PointCloud *pcl) const {
+	if (!pcl->HasNormals() && params_.scanMatcher_.icpObjective_ == o3d_slam::IcpObjective::PointToPlane) {
+		estimateNormals(params_.scanMatcher_.kNNnormalEstimation_, pcl);
```

Comment:
Side question: Do we care about surface normal orientation? That consistency might help in other applications such as mesh generation or gradient detection

Reply:
Yes, but this can also be reoriented right before you use it. It's just one function call.